### PR TITLE
sd-varlink: add function to configure server object info

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 1]
```diff
@@ -1192,35 +1191,46 @@ static int generic_method_get_info(
...
+        if (r < 0)
+                return r;
+
+        r = sd_json_variant_set_field_string(&ret, "version", link->server->version);
+        if (r < 0)
+                return r;
+
+        r = sd_json_variant_set_field_string(&ret, "url", link->server->url);
+        if (r < 0)
+                return r;
```
Comment (YHNdnzj) [inline src/libsystemd/sd-varlink/sd-varlink.c]: Hmm, I'd still use sd_varlink_replybo for these...

File: src/shared/varlink-systemd.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "string-util.h"
+#include "varlink-systemd.h"
+#include "version.h"
+
+int varlink_set_info_systemd(sd_varlink_server *server) {
```
Comment (YHNdnzj) [inline src/shared/varlink-systemd.c]: Why not place this in `varlink-util`? `-util.[ch]` is where helpers for internal use usually lies? E.g. `bus-util`, `device-util`, `json-util`, etc.

Comment (yuwata) [inline src/shared/varlink-systemd.c]: I'd like to make this also internally call `sd_varlink_server_new()`, and rename this to e.g. `varlink_server_new()`, `varlink_server_new_systemd()`, `varlink_server_new_default()`, or so.

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 3]
```diff
@@ -3262,12 +3272,45 @@ static sd_varlink_server* varlink_server_destroy(sd_varlink_server *s) {
...
+                *b = strdup(product),
+                *c = strdup(version),
+                *d = strdup(url);
+        if (!a || !b || !c || !d)
+                return log_oom_debug();
+
+        free_and_replace(s->vendor, a);
+        free_and_replace(s->product, b);
+        free_and_replace(s->version, c);
+        free_and_replace(s->url, d);
```
Comment (yuwata) [inline src/libsystemd/sd-varlink/sd-varlink.c:3290]: Why not to use `free_and_strdup()` ? Then, this can accept `NULL`.

Comment (poettering) [reply src/libsystemd/sd-varlink/sd-varlink.c:3290]: yeah, i think too that we should allow nullifying fields

Comment (keszybz) [reply src/libsystemd/sd-varlink/sd-varlink.c:3290]: Then it wouldn't be atomic on failure. But you're right, it should accept NULL as arguments.

**Comment by keszybz** [comment]
Updated.

File: src/libsystemd/sd-varlink/sd-varlink.c [hunk 4]
```diff
@@ -3262,12 +3257,41 @@ static sd_varlink_server* varlink_server_destroy(sd_varlink_server *s) {
...
+                const char *url) {
+
+        assert_return(s, -EINVAL);
+
+        _cleanup_free_ char
+                *a = vendor ? strdup(vendor) : NULL,
+                *b = product ? strdup(product) : NULL,
+                *c = version ? strdup(version) : NULL,
+                *d = url ? strdup(url) : NULL;
+        if ((vendor && !a) || (product && !b) || (version && !c) || (url && !d))
```
Comment (yuwata) [inline src/libsystemd/sd-varlink/sd-varlink.c:3284]: Maybe better to use `strdup_to()`.

Comment (keszybz) [reply src/libsystemd/sd-varlink/sd-varlink.c:3284]: It comes out to about the same level of complication. We'd need to either collect the return values or use multiple checks. I think the current version is good enough…

Comment (yuwata) [reply src/libsystemd/sd-varlink/sd-varlink.c:3284]: I do not think so, but at the same time, not a strong opinion. So, feel free to ignore.

**Comment by bluca** [comment]
Looks new but unrelated:

```
systemd-journald[776]: ../src/src/shared/copy.c:463:37: runtime error: variable length array bound evaluates to non-positive value 0
```

**Pull request merged**